### PR TITLE
Continued improvements to registration machinery

### PR DESCRIPTION
**Pull request by lavalamp**
---

This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/31399)

### CODE REVIEW

File: pkg/apis/batch/v2alpha1/register.go [hunk 1]
```diff
@@ -17,34 +17,41 @@ limitations under the License.
...
-		&JobList{},
-		&JobTemplate{},
-		&ScheduledJob{},
-		&ScheduledJobList{},
-		&v1.ListOptions{},
-		&v1.DeleteOptions{},
-	)
-	versionedwatch.AddToGroupVersion(scheme, SchemeGroupVersion)
-	return nil
+func init() {
```
Comment (luxas) [inline pkg/apis/batch/v2alpha1/register.go]: just please don't use `func init()` and `import _ "foo"`! It adds SO much overhead when using hyperkube for example.
I'm gonna send a PR soon that removes the most of them, I think it's gonna fix up arm hyperkube

**Comment by luxas** [comment]
Is this going to make v1.4? I hope so, as it would be easier to remove the inits then

File: pkg/apis/batch/v2alpha1/register.go (hunk 1 above)
Comment (lavalamp) [inline pkg/apis/batch/v2alpha1/register.go]: That's coming in the future. The goal here was to make 0 effective change so I don't have to worry about breaking things. Then the PR that actually removes the init statements will be small and easy to reason about. Hopefully!

**Comment by lavalamp** [comment]
Rebased. (easy, fortunately)

**Comment by lavalamp** [comment]
```
2016/09/07 10:18:34 Failed executing generator: some packages had errors:
type k8s.io/kubernetes/pkg/apimachinery/announced.GroupVersionFactoryArgs cannot be converted to protobuf: unable to embed type "[]k8s.io/kubernetes/pkg/runtime.Object" as field "" in "k8s.io/kubernetes/pkg/apimachinery/announced.GroupVersionFactoryArgs": did not recognize the provided type
```

@smarterclayton any idea why proto gen is acting on this package all of a sudden?

**Comment by smarterclayton** [comment]
Might have been included during recursive descent for another type?
Generation is opt out for the package so maybe it got linked through
another top level Struct.

**Comment by lavalamp** [comment]
Yeah. It's just weird that it only started now. I'll investigate more after #31731 lands.

**Comment by luxas** [comment]
@lavalamp Is this something you're gonna continue working on or should we close it?
Or have you just moved this code to other PRs?

**Comment by lavalamp** [comment]
I am not done with this but it may be a while before I have time to focus
on it :(

**Comment by lavalamp** [comment]
sigh, some day I'll get back to this

**Pull request closed without merging**